### ISSUE TITLE
Ruby

### ISSUE THREAD

**Issue by acm-wq** [catppuccin/catppuccin#2556]
### Link to repository

https://github.com/acm-wq/catppuccin_ruby

### Screenshots

![photo_2024-09-22_10-00-29](https://github.com/user-attachments/assets/003a753a-108d-4bd3-a044-9989e3e768c7)
![photo_2024-09-22_10-00-26](https://github.com/user-attachments/assets/fb8ca691-8fbd-4aff-8568-e537a5276209)
![photo_2024-09-22_10-00-16](https://github.com/user-attachments/assets/0f17ec3e-398f-4e30-a5b1-16bee705b307)

### Any additional comments?

I put a lot of effort into creating this port of the Catppuccin color palette for Ruby, and I'd be very happy if you like it!

### Submission Guidelines

- [X] I have read and followed Catppuccin's submission guidelines.

**Comment by backwardspy** [comment]
nice work!

some of our other code libraries such as [rust](https://github.com/catppuccin/rust/blob/main/build.rs) and [python](https://github.com/catppuccin/python/blob/main/build.py) are generated automatically from the [palette json file](https://github.com/catppuccin/palette/blob/main/palette.json). we can also include [a workflow](https://github.com/catppuccin/rust/blob/main/.github/workflows/palette.yml) to automatically update the palette when it changes.

do you think this is feasible for ruby too?

**Comment by acm-wq** [comment]
Thank you! 

Yes, automatically generating Ruby code from a JSON palette file is quite possible.

**Comment by sgoudham** [comment]
> Yes, automatically generating Ruby code from a JSON palette file is quite possible.

Are you interested in taking on this work? It'd be nice to have it implemented before its transferred to the organisation but I don't want to make it a strict requirement since I think that may lead this issue going stale.

**Comment by acm-wq** [comment]
Yeah, I think I can do that. I'd be happy to help!

**Comment by acm-wq** [comment]
> > Yes, automatically generating Ruby code from a JSON palette file is quite possible.
> 
> Are you interested in taking on this work? It'd be nice to have it implemented before its transferred to the organisation but I don't want to make it a strict requirement since I think that may lead this issue going stale.

Hi, I've updated the repository and added generation from json file. Hopefully this is exactly what was needed, but in any case I'm willing to make changes if needed!

**Comment by sgoudham** [comment]
Hi :wave:,
Sorry for the wait on this.

Thanks for including a method of generation from the palette.json file. However, I still think modifications need to be made. If I'm reading [this](https://github.com/acm-wq/catppuccin_ruby/blob/01c9183b621933171016947a4766ce30b72fed2b/lib/catppuccin_ruby.rb#L17-L18) correctly, the user needs the `palette.json` in their local directory in order for the Ruby theme to be intialised.

What I was trying to convey is that the entire Ruby file should be "generated" in some way from the palette.json. I would imagine that this would require another file to load the json file, convert it to Ruby syntax and write the code into a "catppuccin_ruby.rb` file and only that file is shipped to the user. This means that the user doesn't need to install any other dependencies or need the palette.json. However, feel free to correct me if I have read the code wrong as I don't really write Ruby.

Another thing I've noticed is that I see that the gem is currently present on RubyGems. We like to have all our libraries under shared organisation accounts as it stops libraries from being unpublishable if the maintainer moves on. Can you please add the official [catppuccin](https://rubygems.org/profiles/catppuccin) account (`release@catppuccin.com`) as an owner of the package? I believe there are instructions for that in "[Managing Owners in UI - RubyGems](https://guides.rubygems.org/managing-owners-using-ui/)."

Thank you for being patient with us on this one!

**Comment by acm-wq** [comment]
> Hi 👋, Sorry for the wait on this.
> 
> Thanks for including a method of generation from the palette.json file. However, I still think modifications need to be made. If I'm reading [this](https://github.com/acm-wq/catppuccin_ruby/blob/01c9183b621933171016947a4766ce30b72fed2b/lib/catppuccin_ruby.rb#L17-L18) correctly, the user needs the `palette.json` in their local directory in order for the Ruby theme to be intialised.
> 
> What I was trying to convey is that the entire Ruby file should be "generated" in some way from the palette.json. I would imagine that this would require another Ruby file to load the palette.json file, convert it to Ruby syntax and write the generated Ruby code into a "catppuccin_ruby.rb` file and only that file is shipped to the user. This means that the user doesn't need to install any other dependencies or need the palette.json. However, feel free to correct me if I have read the code wrong as I don't really write Ruby.
> 
> Another thing I've noticed is that I see that the gem is currently present on RubyGems. We like to have all our libraries under shared organisation accounts as it stops libraries from being unpublishable if the maintainer moves on. Can you please add the official [catppuccin](https://rubygems.org/profiles/catppuccin) account (`releases@catppuccin.com`) as an owner of the package? I believe there are instructions for that in "[Managing Owners Using UI - RubyGems](https://guides.rubygems.org/managing-owners-using-ui/)."
> 
> Thank you for being patient with us on this one!

Hello, please forgive the long reply! As far as I understand, you want to make two script files that when run will create a full main file with pallitra and methods?

And about the gem, I'll give you the owner soon. Thanks!!!

**Comment by sgoudham** [comment]
> Hello, please forgive the long reply! As far as I understand, you want to make two script files that when run will create a full main file with pallitra and methods?

No worries! Ultimately yes, I think you would have the following repository structure:

```
.
├── LICENSE
├── README.md
├── catppuccin.gemspec
└── lib
    ├── catppuccin.rb
    ├── generate_catppuccin.rb
    └── palette.json
```

I would expect the `generate_catppuccin.rb` file to read the `palette.json` file and _generate_ all the Ruby code to be placed within `catppuccin.rb`. This means that `catppuccin.rb` doesn't need any testing as the hex colours and various other formats are derived from the `palette.json` file, also it means that users aren't loading the `palette.json` themselves at runtime (which I think is what your current code is doing)

And then I think the `catppuccin.gemspec` file would look like this:

```rb
Gem::Specification.new do |s|
  s.name        = 'catppuccin'
  s.version     = '0.1.0'
  s.date        = '2024-09-22'
  s.summary     = "Soothing pastel library for Ruby"
  s.description = "A Ruby gem providing the Catppuccin color palette."
  s.authors     = ["Catppuccin Org", "acm-wq"]
  s.email       = ["releases@catppuccin.com", "camisglhdadae@gmail.com"]
  s.files       = ["lib/catppuccin.rb"]
  s.homepage    = 'https://github.com/catppuccin/ruby'
  s.license     = 'MIT'
end
```

I hope that makes it a bit more clear?

> And about the gem, I'll give you the owner soon. Thanks!!!

I saw that you gave the organisation account owner, thank you!

**Comment by sgoudham** [comment]
I, with the help @backwardspy, also noticed some interesting design decisions:

1. The [`get_random_color()`](https://github.com/acm-wq/catppuccin_ruby/blob/01c9183b621933171016947a4766ce30b72fed2b/lib/catppuccin_ruby.rb#L121-L158) function seems unnecessary and it's not something that our other libraries offer as part of its API. This should be removed.
2. It feels odd that we're always returning strings back to the user instead of the raw object. Returning strings, especially from a colour palette, from a library is often limiting and frustrating instead of being given the raw object to modify/manipulate yourself to display however you wish to do so. I think it would be good to redesign the functions to stop returning strings.

**Comment by acm-wq** [comment]
> I saw that you gave the organisation account owner, thank you!

Yes, thanks for the detailed reply!  
As for the random color feature, I will also redo that as well

**Comment by acm-wq** [comment]
@sgoudham 
Hi! I finished the job. All changes have been committed and sent to the repository. Thanks for your suggestions and feedback!

**Comment by sgoudham** [comment]
Thank you so much @acm-wq!

The new structure looks great! I think we can improve it in the future by adding the newly added ANSI colours but I don't want to block the transfer based on that. I'm happy for it to be transferred over to the organisation! 

Please read, ["Transferring a repository owned by your personal account"](https://docs.github.com/en/repositories/creating-and-managing-repositories/transferring-a-repository#transferring-a-repository-owned-by-your-personal-account), if you are unsure about the process of transferring the repository to Catppuccin.

**We'll close this issue once the port has been added to our [ports.yml](https://github.com/catppuccin/catppuccin/blob/main/resources/ports.yml)**.

**Comment by acm-wq** [comment]
Everything seems to have worked out!
